README.adoc
= WiFi library for the Arduino WiFi Shield 101 and MKR1000 board =

image:https://travis-ci.org/arduino-libraries/WiFi101.svg?branch=master["Build Status", link="https://travis-ci.org/arduino-libraries/WiFi101"]

This library implements a network driver for devices based
on the ATMEL WINC1500 WiFi module.

For more information about this library please visit us at
https://www.arduino.cc/en/Reference/WiFi101

== License ==

Copyright (c) Arduino LLC. All right reserved.

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA


CHANGELOG
WiFi101 ?.?.? - ????.??.??

WiFi101 0.12.1 - 2017.01.19

* Fixed tomorrow day issue in WiFi.getTime()

WiFi101 0.12.0 - 2017.01.05

* Made provisioning mode easier and added example sketch
* Fixed WiFi.getTime() not returning 0, if time has not been synced via NTP
* Fixed crashing when connecting after scanning
* Fixed WiFiServer::available() returning valid client on outbound client connection
* Added WiFiUdp::beginMulticast(port) function for compatiblity with EthernetUdp, as beginMulti was inconsistent

WiFi101 0.11.2 - 2016.12.15

* Fixed value of WiFi.getTime() being off by one day
* Fixed calling WiFi.RSSI() causing lockups

WiFi101 0.11.1 - 2016.11.29

* Fixed regression for non-AVR boards that resulted in corrupt with data over than 1400 bytes was received

WiFi101 0.11.0 - 2016.11.14

* Changed WiFi.ping(...) to return round trip time on success, negative value on error. Thanks @PKGeorgiev
* WiFi.end() now powers down the WiFi module
* WiFi.config(ip) can now be used to set a static IP for WiFi.begin(...)
* Fixed WiFi.BSSID(bssid) returning reversed MAC address
* Added WiFi.APClientMacAddress() API to get MAC address of AP client in AP mode
* Added WiFi.getTime() API to get epoch from NTP

WiFi101 0.10.0 - 2016.09.08

* Added WiFi.end() to disconnect from the AP or end AP mode
* Added new WiFi.ping(...) functionality. Thanks @PKGeorgiev
* Added WiFi.setPins(...) to customize the CS, INTN, RESET and CHIPEN pins 
* Add new WL_AP_LISTENING, WL_AP_CONNECTED, and WL_AP_FAILED status types for AP mode
* Fixed return value of WiFiUDP::beginPacket(host, port) when host is successfully resolved
* Added power management methods: WiFi.lowPowerMode(), WiFi.maxLowPowerMode(), WiFi.noLowPowerMode()
* Close TCP sockets when physical link is disconnected
* Fixed WiFi.RSSI() returning 0 when there was pending socket data

WiFi101 0.9.1 - 2016.04.19

* Increased compatibility with 3rd party boards and architectures.
* Rename WiFiMdnsResponder.h/cpp to WiFiMDNSResponder.h/cpp
* Fixed buffering of UDP packet

WiFi101 0.9.0 - 2016.03.21

* Fixed WiFi.status() hanging when shield is not present
* Fixed MAC address returning 00:00:00:00:00:00 with firmware version 19.3.0
* Fixed SSL write functionality with firmware version 19.3.0
* Fixed previous version of the library not working on the Uno and other boards that used PCINT's
* Added beginAP(...) API's to create access point with WEP passwords
* Fixed beginAP(...) channel off by one error
* Fixed WiFi.status() always returning WL_CONNECTED once connected, even if access point is disconnected later
* Added beginMulti API for Multicast UDP support
* Added WiFiMDNSResponder class and MDNS Web server example

WiFi101 0.8.0 - 2016.02.15

* Added example for Access Point web server (thanks @ladyada)
* Fixed MAC Address printed in reverse order
* Allow another library to override PCINT ISR (fix issues when using the WiFi101
  library with other libraries like SoftwareSerial)

WiFi101 0.7.0 - 2015.01.11

* Added support for WiFi Firmware 19.4.4
* WiFi.hostByName(...) will not try resolve the domain name if it's already a numeric IP
* Fixed manual IP configuration (no DHCP)
* Fixed WiFiServer.available(), now follows API specification
* Fixed WEP key connection
* Fixed WiFiClient copy constructor and assignment operator. This improves stability when
  Client objects are assigned or returned from functions.
* Control pins are now configurable through defines from variant.

WiFi101 0.6.0 - 2015.11.27

* Fixed bug with AVR boards when Web Server is used
* Fixed UDP read bug on AVR Boards
* Added missing inlcude for SSL Client
* Fixed peek() function
* Fixed some examples

WiFi101 0.5.1 - 2015.10.06

* Improved support for AVR Boards (Uno, Mega, Leonardo, etc.) and
  ARM based boards (Due and Zero).

WiFi101 0.5.0 - 2015.10.01

* Initial release



src/WiFi101.h
/*
  WiFi.h - Library for Arduino Wifi shield.
  Copyright (c) 2011-2014 Arduino.  All right reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef WIFI_H
#define WIFI_H

#define WIFI_FIRMWARE_REQUIRED "19.4.4"

#include <Arduino.h>

extern "C" {
#include "driver/include/m2m_wifi.h"
#include "socket/include/socket.h"
}

#include "WiFiClient.h"
#include "WiFiSSLClient.h"
#include "WiFiServer.h"

typedef enum {
	WL_NO_SHIELD = 255,
	WL_IDLE_STATUS = 0,
	WL_NO_SSID_AVAIL,
	WL_SCAN_COMPLETED,
	WL_CONNECTED,
	WL_CONNECT_FAILED,
	WL_CONNECTION_LOST,
	WL_DISCONNECTED,
	WL_AP_LISTENING,
	WL_AP_CONNECTED,
	WL_AP_FAILED,
	WL_PROVISIONING,
	WL_PROVISIONING_FAILED
} wl_status_t;

/* Encryption modes */
enum wl_enc_type {  /* Values map to 802.11 encryption suites... */
	ENC_TYPE_WEP = M2M_WIFI_SEC_WEP,
	ENC_TYPE_TKIP = M2M_WIFI_SEC_WPA_PSK,
	ENC_TYPE_CCMP = M2M_WIFI_SEC_802_1X,
	/* ... except these two, 7 and 8 are reserved in 802.11-2007 */
	ENC_TYPE_NONE = M2M_WIFI_SEC_OPEN,
	ENC_TYPE_AUTO = M2M_WIFI_SEC_INVALID
};

typedef enum {
	WL_RESET_MODE = 0,
	WL_STA_MODE,
	WL_PROV_MODE,
	WL_AP_MODE
} wl_mode_t;

typedef enum {
	WL_PING_DEST_UNREACHABLE = -1,
	WL_PING_TIMEOUT = -2,
	WL_PING_UNKNOWN_HOST = -3,
	WL_PING_ERROR = -4
} wl_ping_result_t;

typedef void(*TGAM_StatusChangeEventCallback) (const wl_status_t newStatus);

class WiFiClass
{
public:
	uint32_t _localip;
	uint32_t _submask;
	uint32_t _gateway;
	int _dhcp;
	uint32_t _resolve;
	byte *_remoteMacAddress;
	wl_mode_t _mode;
	wl_status_t _status;
	char _scan_ssid[M2M_MAX_SSID_LEN];
	uint8_t _scan_auth;
	char _ssid[M2M_MAX_SSID_LEN];
	WiFiClient *_client[TCP_SOCK_MAX];
	//Pho
	static TGAM_StatusChangeEventCallback _globalStatusChangedCallback;
	void setStatusChangedCallback(TGAM_StatusChangeEventCallback response);
	bool _hasAttachedStatusChangeEventCallback;
	// Wait for connection or timeout:
	unsigned long connection_begin_start;
	char connection_begin_ssid[M2M_MAX_SSID_LEN];

	WiFiClass();

	void setPins(int8_t cs, int8_t irq, int8_t rst, int8_t en = -1);

	int init();

	char* firmwareVersion();

	/* Start Wifi connection with WPA/WPA2 encryption.
	 *
	 * param ssid: Pointer to the SSID string.
	 * param key: Key input buffer.
	 */
	uint8_t begin();
	uint8_t begin(const char *ssid);
	uint8_t begin(const char *ssid, uint8_t key_idx, const char* key);
	uint8_t begin(const char *ssid, const char *key);

	uint8_t begin(const String &ssid) { return begin(ssid.c_str()); }
	uint8_t begin(const String &ssid, uint8_t key_idx, const String &key) { return begin(ssid.c_str(), key_idx, key.c_str()); }
	uint8_t begin(const String &ssid, const String &key) { return begin(ssid.c_str(), key.c_str()); }

	bool beginAsync(const char *ssid, const char *key, TGAM_StatusChangeEventCallback callback);
	bool beginAsync(const String &ssid, const String &key, TGAM_StatusChangeEventCallback callback) { return beginAsync(ssid.c_str(), key.c_str(), callback); }
	bool updateConnect();

	/* Start Wifi in Access Point, with open security.
	 * Only one client can connect to the AP at a time.
	 *
	 * param ssid: Pointer to the SSID string.
	 * param channel: Wifi channel to use. Valid values are 1-12.
	 */
	uint8_t beginAP(const char *ssid);
	uint8_t beginAP(const char *ssid, uint8_t channel);
	uint8_t beginAP(const char *ssid, uint8_t key_idx, const char* key);
	uint8_t beginAP(const char *ssid, uint8_t key_idx, const char* key, uint8_t channel);

	uint8_t beginProvision();
	uint8_t beginProvision(uint8_t channel);

	uint32_t provisioned();

	void config(IPAddress local_ip);
	void config(IPAddress local_ip, IPAddress dns_server);
	void config(IPAddress local_ip, IPAddress dns_server, IPAddress gateway);
	void config(IPAddress local_ip, IPAddress dns_server, IPAddress gateway, IPAddress subnet);

	void disconnect();
	void end();

	uint8_t *macAddress(uint8_t *mac);

	uint32_t localIP();
	uint32_t subnetMask();
	uint32_t gatewayIP();
	char* SSID();
	int32_t RSSI();
	uint8_t encryptionType();
	uint8_t* BSSID(uint8_t* bssid);
	uint8_t* APClientMacAddress(uint8_t* mac);
	int8_t scanNetworks();
	char* SSID(uint8_t pos);
	int32_t RSSI(uint8_t pos);
	uint8_t encryptionType(uint8_t pos);

	uint8_t status();

	int hostByName(const char* hostname, IPAddress& result);
	int hostByName(const String &hostname, IPAddress& result) { return hostByName(hostname.c_str(), result); }

	int ping(const char* hostname, uint8_t ttl = 128);
	int ping(const String &hostname, uint8_t ttl = 128);
	int ping(IPAddress host, uint8_t ttl = 128);

	unsigned long getTime();

	void refresh(void);

	void lowPowerMode(void);
	void maxLowPowerMode(void);
	void noLowPowerMode(void);

private:
	int _init;
	char _version[9];

	uint8_t startConnect(const char *ssid, uint8_t u8SecType, const void *pvAuthInfo);
	bool startConnect(const char * ssid, uint8_t u8SecType, const void * pvAuthInfo, TGAM_StatusChangeEventCallback callback);
	wl_status_t completeStartConnect();

	uint8_t startAP(const char *ssid, uint8_t u8SecType, const void *pvAuthInfo, uint8_t channel);
	uint8_t* remoteMacAddress(uint8_t* remoteMacAddress);

	uint8_t startProvision(const char *ssid, const char *url, uint8_t channel);
};

extern WiFiClass WiFi;

#endif /* WIFI_H */
